Snowplow/SPUtilities.m

```diff
@@ -90,17 +88,33 @@ + (NSString *) getOpenIdfa {
     return idfa;
 }
 
+/*
+ The IDFA can be retrieved using selectors rather than proper instance methods because
+ the compiler would complain about the missing AdSupport framework.
+ As stated in the header file, this only works if you have the AdSupport library in your project.
+ If you have it, but do not want to use IDFA, add the compiler flag <code>SNOWPLOW_NO_IFA</code> to your build settings.
+ If you haven't AdSupport framework in your project it just compiles returning a nil advertisingIdentifier.
+ */
 + (NSString *) getAppleIdfa {
-    NSString* idfa = nil;
 #if SNOWPLOW_TARGET_IOS || SNOWPLOW_TARGET_TV
 #ifndef SNOWPLOW_NO_IFA
-    if([[ASIdentifierManager sharedManager] isAdvertisingTrackingEnabled]) {
-        NSUUID *identifier = [[ASIdentifierManager sharedManager] advertisingIdentifier];
-        idfa = [identifier UUIDString];
-    }
+    Class ASIdentifierManagerClass = NSClassFromString(@"ASIdentifierManager");
+    if (!ASIdentifierManagerClass) return nil;
+
+    SEL sharedManagerSelector = NSSelectorFromString(@"sharedManager");
+    id sharedManager = ((id (*)(id, SEL))[ASIdentifierManagerClass methodForSelector:sharedManagerSelector])(ASIdentifierManagerClass, sharedManagerSelector);
```

Comment:
As we're now using reflection to call these methods, I wonder if we should code more defensively. This works fine in the current release of iOS but if Apple change the method names in the future this will cause a hard crash.
Before using the SEL we could call:
`if (![ASIdentifierManagerClass respondsToSelector:sharedManagerSelector]) return nil;`

At least this way, rather than a hard crash occuring we will just fail to track IDFA. I think we should add this to the three uses of NSSelectorFromString, thoughts?

Reply:
👍 It seems a good idea